src/libstd/io/mod.rs

```diff
@@ -867,6 +884,85 @@ pub trait Read {
     }
 }
 
+/// A buffer type used with `Read::read_vectored`.
+///
+/// It is semantically a wrapper around an `&mut [u8]`, but is guaranteed to be
+/// ABI compatible with the `iovec` type on Unix platforms and `WSABUF` on
+/// Windows.
+#[unstable(feature = "iovec", issue = "0")]
+#[repr(transparent)]
+pub struct IoVecMut<'a>(sys::io::IoVecMut<'a>);
+
+#[unstable(feature = "iovec", issue = "0")]
+impl<'a> fmt::Debug for IoVecMut<'a> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(self.as_slice(), fmt)
+    }
+}
+
+impl<'a> IoVecMut<'a> {
+    /// Creates a new `IoVecMut` wrapping a byte slice.
+    ///
+    /// # Panics
+    ///
+    /// Panics on Windows if the slice is larger than 4GB.
+    #[unstable(feature = "iovec", issue = "0")]
+    #[inline]
+    pub fn new(buf: &'a mut [u8]) -> IoVecMut<'a> {
+        IoVecMut(sys::io::IoVecMut::new(buf))
+    }
+
+    /// Returns a shared reference to the inner slice.
+    #[unstable(feature = "iovec", issue = "0")]
+    #[inline]
+    pub fn as_slice(&self) -> &'a [u8] {
+        self.0.as_slice()
+    }
+
+    /// Returns a mutable reference to the inner slice.
+    #[unstable(feature = "iovec", issue = "0")]
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &'a mut [u8] {
```

Comment:
I think the `'a` on the return value here may be unsound?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5780848a8a750eafedf6079de91938ac

Reply:
Ah right. Should we drop the `'a` out of both `as_slice()` and `as_mut_slice()` or just the one?

Reply:
I think both, although the wasm fallback and others is a double-check to ensure the safe implementation works or doesn't.